ocaml-lsp-server/src/ocaml_lsp_server.ml

```diff
@@ -164,22 +169,30 @@ let on_initialize rpc (ip : Lsp.Types.InitializeParams.t) =
   Fiber.return @@ Ok (initialize_info, state)
 
 let code_action server (params : CodeActionParams.t) =
+  let open Fiber.Result.O in
   let state : State.t = Server.state server in
   let store = state.store in
-  match params.context.only with
-  | Some set
-    when not (List.mem (CodeActionKind.Other Destruct_lsp.action_kind) ~set) ->
-    Fiber.return (Ok (None, state))
-  | Some _
-  | None ->
-    let open Fiber.Result.O in
-    let uri = Uri.t_of_yojson (`String params.textDocument.uri) in
-    let* doc = Fiber.return (Document_store.get store uri) in
-    let+ action = Destruct_lsp.code_action doc params in
-    let action =
-      Option.map action ~f:(fun destruct -> [ `CodeAction destruct ])
-    in
-    (action, state)
+  let uri = Uri.t_of_yojson (`String params.textDocument.uri) in
+  let* doc = Fiber.return (Document_store.get store uri) in
+  let code_action kind f =
+    match params.context.only with
+    | Some set when not (List.mem kind ~set) -> Fiber.return (Ok None)
+    | Some _
+    | None ->
+      let+ action = f () in
+      Option.map action ~f:(fun destruct -> `CodeAction destruct)
+  in
+  let* destruct_action =
+    code_action (CodeActionKind.Other Destruct_lsp.action_kind) (fun () ->
+        Destruct_lsp.code_action doc params)
+  in
+  let* inferred_intf_action =
```

Comment:
You should be able to generate these actions concurrently with `Fiber.parallel_map`. It doesn't matter now, but hopefully it will once we add non merlin based actions.